_posts/blog/2016-11-09-immutable-bytes.md

```diff
@@ -0,0 +1,180 @@
+Changing the position conceptually changes the contents of the ByteBuffer.  This is why `hashCode()`
+returns something different after `bb1.get()` is called.  So even though `bb1` does not enable
+mutating `bytes1`, `bb1` is itself mutable.  One might think that calling `map.put(bb1.duplicate(),
+aValue)` would avoid issues.  However, any code iterating over a maps keys could mutate the
+ByteBuffers position.
+
+## Why not use Protobufs ByteString?
+
+[Protocol Buffers][pb] has a beautiful implementation of an immutable byte array wrapper called
+[ByteString].  I would encourage its use when possible.  However in Fluo's case its not really
+appropriate to use for two reasons.  First any library designer should try to minimize what
+transitive dependencies they force on users.  Internally Fluo does not currently use Protocol
+Buffers in its implementation, so this would be a new dependency for Fluo users.  The second reason
+is going to require some background to explain.
+
+Technologies like [OSGI] and [Jigsaw] seek to modularize Java libraries and provide dependency
+isolation.  Dependency isolation allows a user to use a library without having to share a libraries
+dependencies.  For example, consider the following hypothetical scenario.
+
+ * Fluo's implementation uses Protobuf version 2.5
+ * Fluo user code uses Protobuf version 1.8
+
+Without dependency isolation the user must converge dependencies and make their application and
+Fluo's implementation use the same version of Protobuf.  Sometimes this works without issue, but
+sometimes things will break because Protobuf dropped, changed, or added a method.
+
+With dependency isolation, Fluo's implementation and Fluo user code can easily use different versions
+of Protobuf.  This is only true as long as Fluo's API does not use Protobuf.  So this the second
+reason that Fluo should not use classes from Protobuf in its API.  If Fluo used Protobuf in its API
+then it forces the user to have to converge dependencies, even if they are using OSGI or Jigsaw. 
+
+## What about the copies?
+
+As mentioned earlier, an Immutable type requires a defensive copy at creation time.  When we were
+designing Fluo's API we were worried about this at first.  However a simple truth became apparent.
+If the API took a mutable type, then all boundary points between the user and Fluo would require
+defensive copies.  For example assume Fluo's API took byte arrays and consider the following code.
+
+```java
+//A Fluo transaction
+Transaction tx = ...
+byte[] row = ...
+
+tx.set(row, col1, v1)
+tx.set(row, col2, v2)
+tx.set(row, col3, v3)
+```
+
+Fluo will buffer changes until a transaction is committed.  In the example above since Fluo accepts
+a mutable row, it would be prudent to do a defensive copy each time set is called above.  
+
+In the code below where an immutable byte array wrapper is used, the calls to set do not need to do
```

Comment:
Put set as `set` so it's clear you're using the method name, and not the word "set".